chainer/functions/loss/discriminative_loss.py

```diff
@@ -35,17 +38,12 @@ class DiscriminativeMarginBasedClusteringLoss(object):
     """
 
     def __init__(self, delta_v=0.5, delta_d=1.5,
-                 max_embedding_dim=10, norm=1,
-                 alpha=1.0, beta=1.0, gamma=0.001):
+                 norm=1, alpha=1.0, beta=1.0, gamma=0.001):
```

Comment:
Currently all the arguments can be specified as positional. If `max_embedding_dim` is to be removed, all the succeeding arguments must be keyword-only. Otherwise the user codes can be *silently* broken.

You can use `chainer.utils.argument.parse_kwargs` to make them keyword-only
(see `chainer.training.extensions.snapshot()` in `training/extensions/_snapshot.py` as an exapmle).

In that case, please update the docstring to show the "logical" signature.
`chainer.functions.convolution_2d` might be a good example to do that (see the first line of the docstring).

Reply:
There's yet another complexity: deprecation period.

We have a policy to place a deprecation period of at least one major release.
During this period, existing user code with `max_embedding_dim`, or those giving the succeeding arguments as positional, should work just as before (with `DeprecationWarning`).

It can be a bit too tricky to support both the existing and new-style interface at the same time. I think I'd like to do that in a separate PR. In this PR, could you simply leave the argument `max_embedding_dim` and warn if the argument is explicitly given (either as positional or keyword)?

I mean, let the signature as below (note that the default value is `None`):

```py
def __init__(self, delta_v=0.5, delta_d=1.5,
             max_embedding_dim=None, norm=1,
             alpha=1.0, beta=1.0, gamma=0.001):
```

And warn if `max_embedding_dim` is given:

```py
if max_embedding_dim is not None:
    warnings.warn(
        'blah',
        DeprecationWarning)
```

We should also warn if the remaining arguments (`norm`, etc.) are given as positional. I'd like to do that in a separate PR.

Reply:
I have added `**kwargs` option as you suggested. Would it be ok as-is or shall I also add deprecation warning?

Reply:
Thank you for the fix.
It looks the current version misses this point:

>If `max_embedding_dim` is to be removed, all the succeeding arguments must be keyword-only. Otherwise the user codes can be *silently* broken.

For example, if a user has written a code like this:
```
DiscriminativeMarginBasedClusteringLoss(0.5, 1.5, 10, 2)
```

He meant `delta_v=0.5, delta_d=1.5, max_embedding_dim=10, norm=1, alpha=1.0`.
After this fix, this would become `delta_v=0.5, delta_d=1.5, norm=10, alpha=2`, all silently.

We have to keep the behavior of the code like above intact, and at the same time prepare for making all the arguments (`norm`, `alpha`, ...) keyword-only.